server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceReconciler.java

```diff
@@ -561,19 +564,30 @@ private void balance() {
         }
 
         private void maybeLogUndesiredAllocationsWarning(int allAllocations, int undesiredAllocations, int nodeCount) {
-            // more shards than cluster can relocate with one reroute
-            final boolean nonEmptyRelocationBacklog = undesiredAllocations > 2L * nodeCount;
-            final boolean warningThresholdReached = undesiredAllocations > undesiredAllocationsLogThreshold * allAllocations;
-            if (allAllocations > 0 && nonEmptyRelocationBacklog && warningThresholdReached) {
-                undesiredAllocationLogInterval.maybeExecute(
-                    () -> logger.warn(
-                        "[{}] of assigned shards ({}/{}) are not on their desired nodes, which exceeds the warn threshold of [{}]",
+            // There are more shards in undesirable locations than the cluster can relocate with one reroute per shard.
+            final boolean moreThanOneReroutePerNode = undesiredAllocations > 2L * nodeCount;
+            final boolean warningThresholdReached = undesiredAllocations > undesiredAllocationPercentageLogThreshold * allAllocations;
+            if (allAllocations > 0 && moreThanOneReroutePerNode && warningThresholdReached) {
+                undesiredAllocationLogInterval.maybeExecute(() -> {
+                    logger.log(
+                        // The first time a significant imbalance appears, this will log as a warning. But we only want info-level
+                        // logging after that if the undesiredAllocations value moves in a healthy descending direction.
+                        undesiredAllocations > lastNumberOfUndesiredAllocations ? Level.WARN : Level.INFO,
```

Comment:
I wonder if we should only log if it is rising and completely skip (or may be DEBUG) otherwise?

Reply:
I wasn't sure whether it was helpful to see that the value is decreasing in the logs? I remember seeing in a SDHE a comment quoting the log messages showing the decrease over time. Though maybe the lack of the log messages would indicate the same thing if we moved it over to DEBUG only.

Happy to change it to DEBUG if that isn't useful signal.

Reply:
Yeap, fair enough, no message should mean only no warning and not the decreasing value.

Reply:
I am also not sure now if it is worth extra complexity to switch log level to info if the amount is not growing. Why not keep it at warn as it is still not an ideal situation for the cluster?

Reply:
The trouble with using WARN here is that this is not an actionable situation. The user just needs to wait, in most cases, for the cluster to finish rebalancing itself. But the purpose of a WARN message is for the user to do something. So the user is now alarmed and looking for something to do to fix the warning, and comes to us (potentially).